crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//! convert to `isize` because aggregation implicitly multiplies by record
+//! weights):
+//!
+//! ```ignore
+//!         .aggregate_linear(|(_c, _y, _m), &v| v as isize)
+//! ```
+//!
+//! To see our output we can use [`inspect`](`Stream::inspect`) again.  We could
+//! just print the raw (indexed) Z-sets, but here's a slightly refined way that
+//! shows how to break Z-sets down into records:
+//!
+//! ```ignore
+//! monthly_totals.inspect(|data| {
+//!     data.iter()
+//!         .for_each(|((c, y, m), sum, w)| println!("{c:16} {y}-{m:02} {sum:10}: {w:+}"))
+//! });
+//! ```
+//!
+//! The output looks like this:
+//!
+//! ```text
+//! England          2021-01    5600174: +1
+//! England          2021-02    9377418: +1
+//! England          2021-03   11861175: +1
+//! England          2021-04   11288945: +1
+//! England          2021-05   13772946: +1
+//! England          2021-06   10944915: +1
+//! ...
+//! Northern Ireland 2021-01     150315: +1
+//! Northern Ireland 2021-02     317074: +1
+//! ...
+//! Wales            2023-01      33838: +1
+//! Wales            2023-02      17098: +1
+//! Wales            2023-03       8776: +1
+//! ```
+//!
+//! The full program is in `tutorial5.rs`.
+//!
+//! ### Rolling aggregation
+//!
+//! By using a "moving average" to average recent data,
+//! we can obtain a dataset with less noise due to variation from month
+//! to month.   DBSP
+//! provides [`Stream::partitioned_rolling_average`] for this purpose.  To
+//! use it, we have to map our indexed Z-set into a "partitioned indexed
+//! Z-set" ([`PartitionedIndexedZSet`]).
+//! This is just an indexed Z-set in which the key is the "partition"
+//! within which averaging occurs (for us, this is the country), and the
+//! value is a tuple of a "timestamp" and a value.  DBSP uses the
+//! timestamp component, which must have an unsigned integer type, to
+//! define the window.  We can map our `(country, year, month)` key and
+//! `vaccinations` value into a `country` partition, `date` timestamp, and
+//! `vaccinations` value like this:
+//!
+//! ```ignore
+//!     let moving_averages = monthly_totals
+//!         .map_index(|((c, y, m), v)| (c.clone(), (*y as u32 * 12 + (*m as u32 - 1), *v)))
+//! ```
+//!
+//! Once we've done that, computing the moving average is easy.  Here's how we
+//! get the average of the current month and the two preceding months (when
+//! they're in the data set):
+//!
+//! ```ignore
+//!         .partitioned_rolling_average(RelRange::new(RelOffset::Before(2), RelOffset::Before(0)))
+//! ```
+//!
+//! We can re-map to recover year and month from the timestamp that we made.  We
+//! also strip off a `Some(...)` that gets wrapped around the rolling average.
+//! This can end up as a `None` if there are no rows in the window, but, in
+//! turn, that can only happen if the window doesn't include the current row,
+//! and we didn't define our window that way.  Anyhow:
```

Comment:
Maybe worth saying explicitly that `Before(0)` includes the current row.

Reply:
Done.